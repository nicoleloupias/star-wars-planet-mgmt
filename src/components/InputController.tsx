import { FormControl, FormErrorMessage, FormLabel, Input, InputProps } from "@chakra-ui/react";
import { Control, FieldValues, UseControllerProps, useController } from "react-hook-form";

export interface InputControllerProps<T extends FieldValues> extends InputProps {
  id: UseControllerProps<T>["name"];
  control: Control<T>;
  label: string;
  rules?: Record<string, unknown>;
}
export const InputController = <T extends FieldValues>({
  id,
  control,
  label,
  rules,
  ...props
}: InputControllerProps<T>) => {
  const { field, fieldState } = useController<T>({
    name: id,
    control,
    rules
  });

  const errorMessage = fieldState.error?.message;

  return (
    <FormControl isInvalid={fieldState.invalid}>
      <FormLabel htmlFor={id}>{label}</FormLabel>
      <Input data-testid={id} id={id} placeholder={label} {...field} {...props} />
      <FormErrorMessage>{errorMessage}</FormErrorMessage>
    </FormControl>
  );
};
