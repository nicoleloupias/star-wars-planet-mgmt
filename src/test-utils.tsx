import { ChakraProvider } from "@chakra-ui/react";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import React from "react";
import { Router, RouterProvider } from "react-router-dom";
import { theme } from "./theme";
const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <ChakraProvider theme={theme}>
      <Router location={history.location} navigator={history}>
        {children}
      </Router>
    </ChakraProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {}
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => <AllProviders history={history}>{children}</AllProviders>,
    ...options
  });

export { customRender as render };
