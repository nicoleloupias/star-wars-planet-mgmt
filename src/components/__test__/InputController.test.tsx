import { describe, expect, test, vi } from "vitest";
import { render } from "../../test-utils";
import { act, fireEvent, waitFor } from "@testing-library/react";
import type { InputControllerProps } from "../InputController";
import { InputController } from "../InputController";
import type { FieldValues, UseFormReturn } from "react-hook-form";
import { useForm } from "react-hook-form";
type DefaultProps = Omit<InputControllerProps<FieldValues>, "control">;

describe("<InputController/>", () => {
  const defaultProps: DefaultProps = {
    placeholder: "Name",
    rules: { required: "This field is required" },
    id: "name",
    label: "Name"
  };

  let form: UseFormReturn;
  const RenderWithForm = (props: DefaultProps) => {
    const localForm = useForm({
      mode: "onChange",
      criteriaMode: "all"
    });
    form = localForm;
    const { control } = localForm;

    return <InputController control={control} {...props} />;
  };

  const factoryComponent = (props: DefaultProps = defaultProps) => {
    return render(<RenderWithForm {...props} />);
  };

  test("should render correctly", () => {
    const component = factoryComponent();
    const input = component.getByTestId("name");
    expect(input).toBeInTheDocument();
    expect(input).toHaveProperty("placeholder", defaultProps.placeholder);
    expect(component).toMatchSnapshot();
  });

  test("should show error", async () => {
    const component = factoryComponent();
    const input = component.getByTestId("name");

    act(() => {
      fireEvent.change(input, { target: { value: "Alex" } });
    });
    act(() => {
      fireEvent.change(input, { target: { value: "" } });
    });

    await waitFor(() => {
      expect(component.getByText("This field is required")).toBeInTheDocument();
    });

    expect(component).toMatchSnapshot("Invalid");
  });

  test("enables to set errors from outside", async () => {
    const component = factoryComponent();

    act(() => {
      form.setError("name", { type: "invalidName" });
    });

    await waitFor(() => {
      expect(component.getByTestId("name")).toBeInvalid();
    });
  });

  test("should fire custom onChange", async () => {
    const onChange = vi.fn();
    const component = factoryComponent({ ...defaultProps, onChange });
    const inputField = component.getByTestId("name");
    act(() => {
      fireEvent.change(inputField, { target: { value: "Alex" } });
    });

    await waitFor(() => {
      expect(onChange).toHaveBeenCalled();
    });
  });

  test("should fire custom onBlur", async () => {
    const onBlur = vi.fn();
    const component = factoryComponent({ ...defaultProps, onBlur });
    const inputField = component.getByTestId("name");
    act(() => {
      fireEvent.change(inputField, { target: { value: "Alex" } });
      fireEvent.blur(inputField);
    });

    await waitFor(() => {
      expect(onBlur).toHaveBeenCalled();
    });
  });
});
