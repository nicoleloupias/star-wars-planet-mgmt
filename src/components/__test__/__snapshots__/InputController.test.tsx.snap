// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<InputController/> > should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="name"
          id="field-:r0:-label"
        >
          Name
        </label>
        <input
          class="chakra-input css-qcorp"
          data-testid="name"
          id="name"
          name="name"
          placeholder="Name"
          value=""
        />
      </div>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        for="name"
        id="field-:r0:-label"
      >
        Name
      </label>
      <input
        class="chakra-input css-qcorp"
        data-testid="name"
        id="name"
        name="name"
        placeholder="Name"
        value=""
      />
    </div>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<InputController/> > should show error > Invalid 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div>
      <div
        class="chakra-form-control css-1kxonj9"
        data-invalid=""
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          data-invalid=""
          for="name"
          id="field-:r1:-label"
        >
          Name
        </label>
        <input
          aria-describedby="field-:r1:-feedback"
          aria-invalid="true"
          class="chakra-input css-qcorp"
          data-testid="name"
          id="name"
          name="name"
          placeholder="Name"
          value=""
        />
        <div
          aria-live="polite"
          class="chakra-form__error-message css-502kp3"
          id="field-:r1:-feedback"
        >
          This field is required
        </div>
      </div>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-form-control css-1kxonj9"
      data-invalid=""
      role="group"
    >
      <label
        class="chakra-form__label css-g6pte"
        data-invalid=""
        for="name"
        id="field-:r1:-label"
      >
        Name
      </label>
      <input
        aria-describedby="field-:r1:-feedback"
        aria-invalid="true"
        class="chakra-input css-qcorp"
        data-testid="name"
        id="name"
        name="name"
        placeholder="Name"
        value=""
      />
      <div
        aria-live="polite"
        class="chakra-form__error-message css-502kp3"
        id="field-:r1:-feedback"
      >
        This field is required
      </div>
    </div>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
